ledger-service/http-json/src/itlib/scala/http/AbstractWebsocketServiceIntegrationTest.scala

```diff
@@ -747,6 +748,42 @@ abstract class AbstractWebsocketServiceIntegrationTest
     go(createCount)
   }
 
+  /** Updates the ACS retrieved with [[readAcsN]] until the given offset
+    */
+  private[this] def updateAcs(
+      acs: Vector[(String, JsValue)],
+      expectLivenessMarker: Boolean,
+      until: domain.Offset,
+  ): Consume.FCC[JsValue, Map[String, JsValue]] = {
+    val dummyLivenessMarker =
+      JsObject("events" -> JsArray.empty, "offset" -> JsNumber(0)).asInstanceOf[JsValue]
+    val dslSyntax = Consume.syntax[JsValue]
+    import dslSyntax._
+    import domain.Offset.ordering
+    import scalaz.syntax.order._
+    def go(
+        acs: Map[String, JsValue],
+        latest: Option[domain.Offset],
+    ): Consume.FCC[JsValue, Map[String, JsValue]] =
+      if (latest.fold(false)(_ >= until)) {
+        for {
+          ContractDelta(creates, deletes, _) <- readOne
+        } yield acs ++ creates -- deletes.map(_.contractId.unwrap)
+      } else {
+        for {
+          ContractDelta(creates, deletes, offset) <- readOne
+          next <- go(acs ++ creates -- deletes.map(_.contractId.unwrap), offset)
+        } yield next
+      }
+    for {
+      // Ignoring the live marker, w/ per-query offsets it can come in with an offset greater than `until`
```

Comment:
I think it would be nice to have some tests for what we return in the liveness markers. Primarily interested in test case 1 so feel free to omit the others.
1. All offsets come before the ledger end (or rather what the ACS service returns which is technically not guaranteed to be ledger end but that’s what it does atm).
2. All offsets come after the ACS end. You have to "guess" offsets in this case which is a bit odd but not too hard. If the ACS service actually does not return ledger end this matters a bit more.
3. Some mix where a few offsets are before and a few are after.

Reply:
I'll try to find a way to speed up our tests by not reading the heartbeat at the end and include these cases as well.

Reply:
After https://github.com/digital-asset/daml/pull/9847/commits/ab2535a7f091464d915222749bebac2fd0b091d6 tests are down to run in ~1s each, which is reasonable. Will work on this additional test cases next.

Reply:
> All offsets come before the ledger end (or rather what the ACS service returns which is technically not guaranteed to be ledger end but that’s what it does atm).

This is covered by the existing test.

> All offsets come after the ACS end.

_and_

> Some mix where a few offsets are before and a few are after.

What would this test represent? It looks to me like this is not particularly relevant as offsets are opaque to clients and need to be observed before being used (which is a documented property of offsets). I can probably come up with a test for this but I'm not entirely sure of its usefulness.

Reply:
@cocreature I'll mark this as resolved, feel free to reopen if you think there's a good case for these tests to come in.

Reply:
Fine to not add it for now at least. The case I was thinking off is where the ACS service is lagging behind which afaik the docs explicitly state that it can. Our implementation doesn’t do that atm so you can’t hit it but I’m slightly worried this might change and we forget to fix this.